File: android/src/main/java/ai/xmax/reactnative/XmaxImageFrameTask.kt
package ai.xmax.reactnative

/** Image frame metadata, accessed only under the source's shared runtime gate. */
internal class XmaxImageFrameTask {
  private var taskID = ""
  private var index = 0uL

  /** Repeated assignments preserve the sequence; replacing or clearing resets it. */
  fun set(taskID: String) {
    if (this.taskID == taskID) return
    this.taskID = taskID
    index = 0uL
  }

  /** No frame is submitted outside an active generation task. */
  fun nextData(): ByteArray? {
    if (taskID.isEmpty()) return null
    return "$taskID&index=${index++}".toByteArray(Charsets.UTF_8)
  }
}

File: ios/XmaxImageFrameTask.swift
import Foundation

/// Image frame metadata, accessed only under the source's shared runtime gate.
struct XmaxImageFrameTask {
  private var taskID = ""
  private var index: UInt64 = 0

  /// Replaces the task and resets its sequence; repeated assignments are idempotent.
  mutating func set(_ taskID: String) {
    guard self.taskID != taskID else { return }
    self.taskID = taskID
    index = 0
  }

  /// Returns no metadata outside generation, so the source skips that frame.
  mutating func nextData() -> Data? {
    guard !taskID.isEmpty else { return nil }
    let data = Data("\(taskID)&index=\(index)".utf8)
    index &+= 1
    return data
  }
}
